SU2_CFD/src/solver_direct_mean.cpp

```diff
@@ -4744,27 +4754,23 @@ void CEulerSolver::SetTime_Step(CGeometry *geometry, CSolver **solver_container,
   
   /*--- For exact time solution use the minimum delta time of the whole mesh ---*/
   
-  if (time_steping) {
+  if (time_stepping) {
+    /*--- If the unsteady CFL is set to zero, it uses the defined unsteady time step, otherwise
+     * it computes the time step based on the unsteady CFL ---*/
+    if (config->GetUnst_CFL() == 0) {
+      Global_Delta_Time = config->GetDelta_UnstTimeND();
+    } else {
 #ifdef HAVE_MPI
     su2double rbuf_time, sbuf_time;
     sbuf_time = Global_Delta_Time;
     SU2_MPI::Reduce(&sbuf_time, &rbuf_time, 1, MPI_DOUBLE, MPI_MIN, MASTER_NODE, MPI_COMM_WORLD);
     SU2_MPI::Bcast(&rbuf_time, 1, MPI_DOUBLE, MASTER_NODE, MPI_COMM_WORLD);
     Global_Delta_Time = rbuf_time;
 #endif
-    for (iPoint = 0; iPoint < nPointDomain; iPoint++) {
-            
-            /*--- Sets the regular CFL equal to the unsteady CFL ---*/
-            config->SetCFL(iMesh,config->GetUnst_CFL());
-            
-            /*--- If the unsteady CFL is set to zero, it uses the defined unsteady time step, otherwise
-             it computes the time step based on the unsteady CFL ---*/
-            if (config->GetCFL(iMesh) == 0.0) {
-                node[iPoint]->SetDelta_Time(config->GetDelta_UnstTime());
-            } else {
-                node[iPoint]->SetDelta_Time(Global_Delta_Time);
-            }
-        }
+    }
+    for (iPoint = 0; iPoint < nPointDomain; iPoint++)
+      node[iPoint]->SetDelta_Time(Global_Delta_Time);
+    config->AddCurrent_UnstTime(Global_Delta_Time);
```

Comment:
This is okay as long as the value of `Global_Delta_Time` is only used to track how much time has elapsed (i.e., as long as the residual does not depend on time directly).  If the residual depends on time, we would need to make sure the way `Global_Delta_Time` is updated is consistent with the temporal discretization (e.g., that it gets updated properly for RK substeps).  We are not in this case yet, so it is not critical.  But, it should either be fixed or noted somewhere as something to be generalized later.

Reply:
Let me check if I'm understanding what you said:

For explicit time steps, we might want a time step for an internal iteration (a substep).  That means that dt for a substep might _not_ be the same as dt for a full step.  We would want the node `Delta_Time` values to use the proper value (the substep dt).  But here they're being explicitly set to the full timestep (the full dt).

Is that correct?

If that's correct, then we'd have to change how the code works to do that. The time step is only computed once per RK iteration. (See line 163 of `SU2_CFD/src/integration_time.cpp`).  We'd have to change SU2 to compute timesteps at each RK-step or fill in all the substep times once each full step. Neither would be a big problem, but they do represent a change in how SU2 works.

Reply:
No, that isn't what I meant.  Sorry for not being clear.

The issue is not with dt.  There is only a single value of dt for each time step, even for multistage methods, and it remains the same across substeps.  So, the current method of computing dt only on the first RK substep is correct.  However, you do have to be able to evaluate the residual at "intermediate" conditions (state and time) during the substeps.

So, the issue here is that if you actually need to use the value of the time (i.e., the variable `Current_UnstTime` that is updated by the call to `AddCurrent_UnstTime`) when the residual is evaluated (i.e., if you're solving du/dt = f(u,t) rather than du/dt = f(u)), then how that update to the time is done needs to be aware of the temporal discretization.  The current approach of adding dt to  `Current_UnstTime` in `SetTime_Step` is appropriate for backward Euler but not correct for forward Euler or any of the RK schemes.

This doesn't affect us yet since we're solving du/dt = f(u) at the moment (and that won't change unless we do MMS).  So, I'm fine with leaving it as is and putting in a comment and creating a ticket so we don't forget.

Reply:
Alright then. I'll modify the comments I just added to reflect what you just said.